src/common/framework/tree.ts

```diff
@@ -1,82 +1,306 @@
+    if ('children' in tree) {
+      if (tree.description !== undefined) {
+        s += indent + `\n    | ${JSON.stringify(tree.description)}`;
+      }
+
+      for (const [name, child] of tree.children) {
+        s += '\n' + FilterResultTree.subtreeToString(name, child, indent + '  ');
+      }
     }
-    const v = { children: new Map() };
-    children.set(k, v);
-    return v;
+    return s;
   }
+}
 
-  const tree = { children: new Map() };
-  for (const f of listing) {
-    const files = getOrInsert(tree, f.id.suite + ':');
-    if (f.id.path === '') {
-      // This is a suite README.
-      files.description = f.spec.description.trim();
+// TODO: Consider having subqueriesToExpand actually impact the depth-order of params in the tree.
+export async function loadTreeForQuery(
+  loader: TestFileLoader,
+  queryToLoad: TestQuery,
+  subqueriesToExpand: TestQuery[]
+): Promise<FilterResultTree> {
+  const suite = queryToLoad.suite;
+  const specs = await loader.listing(suite);
+
+  const subqueriesToExpandEntries = Array.from(subqueriesToExpand.entries());
+  const seenSubqueriesToExpand: boolean[] = new Array(subqueriesToExpand.length);
+  seenSubqueriesToExpand.fill(false);
+
+  const checkCollapsible = (subquery: TestQuery) =>
+    subqueriesToExpandEntries.every(([i, toExpand]) => {
+      const ordering = compareQueries(toExpand, subquery);
+
+      // If toExpand == subquery, no expansion is needed (but it's still "seen").
+      if (ordering === Ordering.Equal) seenSubqueriesToExpand[i] = true;
+      return ordering !== Ordering.StrictSubset;
+    });
+
+  // L0 = suite-level, e.g. suite:*
+  // L1 =  file-level, e.g. suite:a,b:*
+  // L2 =  test-level, e.g. suite:a,b:c,d:*
+  // L3 =  case-level, e.g. suite:a,b:c,d:
+  let foundCase = false;
+  // L0 is suite:*
+  const subtreeL0 = makeTreeForSuite(suite);
+  checkCollapsible(subtreeL0.query); // mark seenSubqueriesToExpand
+  for (const entry of specs) {
+    if (entry.file.length === 0 && 'readme' in entry) {
+      // Suite-level readme.
+      subtreeL0.description = entry.readme.trim();
       continue;
     }
 
-    let tests = files;
-    for (const path of iteratePath(f.id.path, ':')) {
-      tests = getOrInsert(tests, f.id.suite + ':' + path);
-    }
-    if (f.spec.description) {
-      // This is a directory README or spec file.
-      tests.description = f.spec.description.trim();
+    {
+      const queryL1 = new TestQueryMultiFile(suite, entry.file);
+      const orderingL1 = compareQueries(queryL1, queryToLoad);
+      if (orderingL1 === Ordering.Unordered) {
+        // File path is not matched by this filter.
+        continue;
+      }
     }
 
-    if (!('g' in f.spec)) {
-      // This is a directory README.
+    if ('readme' in entry) {
+      // Entry is a readme that is an ancestor or descendant of the query.
+
+      // readmeSubtree is suite:a,b,*
+      const readmeSubtree: FilterResultSubtree<TestQueryMultiFile> = makeSubtreeForDirPath(
```

Comment:
Why do we need to store READMEs? To surface the description in the standalone runner?

Reply:
Yes. This will always end up being redundant with other tree entries anyway (if there are .spec.ts files anywhere in the directory with the README).